From: SeongJae Park <sj@kernel.org>
Date: Mon, 28 Oct 2024 16:55:18 -0700
Subject: [PATCH] ==== mm/damon: replace most damon_callback usages in sysfs
 with new core functions ====

mm/damon: replace damon_callback usages in sysfs with new core functions

DAMON provides damon_callback API that notifies monitoring events and
allows safe access to damon_ctx internal data.  Users can set and unset
callback functions for different monitoring events in damon_ctx.  Then
DAMON worker thread (kdamond) calls back the functions on the events as
long as the function is set on the damon_ctx.

It is designed in such simple way because it was sufficient for usages
of DAMON at the early days.  We also wanted to make it flexible so that
API client code can implement any required additional features on top of
damon_callback on their demands.

As expected, more sophisticated usages have invented.  Online updates
of parameters or DAMOS auto-tuning inputs and retrieval of DAMOS
statistics and tried regions are notable such usages.  Because
damon_callback doesn't provide any explict synchronization mechanism,
the user ABIs for exposing such functionalities are implemented in
asynchronous ways (DAMON_RECLAIM and DAMON_LRU_SORT}), or synchronous
ways (DAMON_SYSFS) with additional synchronization mechanisms that built
inside the ABI implementation, on top of damon_callback.

So damon_callback is working as expected.  However, the additional
mechanisms built inside ABI on top of damon_callback beocming somewhat
too big and not easy to maintain.  damon_callback is basically simple
enougy to use in not only correct but also wrong ways.  The additional
mechanisms can be easier to maintain when implemented inside core logic.

Introduce two new DAMON core API, namely 'damon_call()' and
'damos_walk()'.  The two functions support synchronous access to
- damon_ctx internal data including DAMON parameters and monitoring
  results, and
- DAMOS-specific data such as regions that each DAMOS action is applied,
respectively.

And replace most of damon_callback usages from DAMON sysfs with the new
core API functions.  Online DAMON parameters tuning feature is not
replaced in this series, since it has specific callback timing
assumptions that require more works.

Patch sequence
==============

First two patches removes unnecessary synchronization dependency of
clear_schemes_tried_regions command handling.

Third patch implements one of the new DAMON API, damon_call().  Three
patches replacing damon_callback usages in DAMON sysfs interface using
damon_call() follow.

Then, seventh patch introduces the other new DAMON API, damos_walk().
Eighth patch replaces two damon_callback usages in DAMON sysfs interface
using damos_walk().  The ninth patch finally cleans up code that no more
being used.

Revision History
================

Changes from RFC
(https://lore.kernel.org/20241213215306.54778-1-sj@kernel.org)
- Call damos_call_cotnrol->walk_fn() after applying action
- Document DAMOS regions walk feature on design doc

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 damon_meta_changes/TYRkgyiG | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 damon_meta_changes/TYRkgyiG

diff --git a/damon_meta_changes/TYRkgyiG b/damon_meta_changes/TYRkgyiG
new file mode 100644
index 000000000000..e69de29bb2d1
-- 
2.39.5

